keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,737 @@
+     specified in `--node-ips`, then we are done.
+
+  6. Otherwise, kubelet will look for an element of the address list
+     that satisfies the first element of `--node-ips`, and (if
+     `--node-ips` has two elements) an element that satisfies the
+     second element of `--node-ips`, and move it/them to the front of
+     the list, in the correct order. It is guaranteed at this point
+     that at least one of those two elements exists, and that moving
+     it/them will result in a list of addresses that matches
+     `--node-ips`.
+
+  7. The resulting list is then assigned to `Node.Status.Addresses`.
+
+Eg, suppose `--node-ips` is `ipv4,ipv6` and the initial list of
+addresses is `[ fd01::1, fd01::2, 10.0.0.1, 10.0.0.2 ]`. In that case
+steps 1-4 would have no effect. Step 5 would determine that the
+Primary Node IP was `fd01::1` and the Secondary Node IP was
+`10.0.0.1`, which does not match `ipv4,ipv6`. (Had we said `ipv6,ipv4`
+instead we'd be done though.) Thus, step 6 would reorder the list to
+`[ 10.0.0.1, fd01::1, fd01::2, 10.0.0.2 ]`, and then the Primary Node
+IP would be `10.0.0.1` and the Secondary would be `fd01::1`, and we
+would match.
+
+With the same list of addresses but `--node-ips 10.0.0.2`, step 2
+would remove the two IPv6 addresses, step 4 would confirm that
+`10.0.0.2` was in the list, step 5 would find the wrong Primary Node
+IP, and step 6 would move the desired IP to the front of the list,
+giving `[ 10.0.0.2, 10.0.0.1 ]`.
+
+With `--node-ips ipv4,ipv6` and initial list of addresses `[ 10.0.0.1,
+10.0.0.2 ]`, steps 1-4 would have no effect, and step 5 would find
+that the Primary Node IP was `10.0.0.1` and there was no Secondary
+Node IP. This _is_ compatible with `--node-ips ipv4,ipv6` (since both
+elements are optional), so we would keep the list of node addresses
+unchanged. Note that we would get exactly the same result with
+`--node-ips ipv6,ipv4`.
+
+```
+<<[UNRESOLVED reordering ]>>
+
+If we wanted to avoid ever rearranging `Node.Status.Addresses`, we
+could add new `NodeAddressType` values `PrimaryIP` and `SecondaryIP`,
+and prepend new elements to the list. This would also allow fix the
+problem where it's currently impossible to make an `ExternalIP` become
+the primary IP.
+
+The down side to this would be that older components using the
+original "Primary Node IP" algorithm would not agree with newer
+clients about what the node's primary IP was.
+
+Perhaps we could handle that by phasing in the new behavior over
+several releases: at first we would add the new addresses _and_ also
+rearrange the old ones, and then eventually we could stop doing the
+rearrangement.
+
+This would also be slightly annoying for clients though because they
+may have to continue dealing with un-updated cloud providers that
+don't set the new fields for a long time.
+
+<<[/UNRESOLVED]>>
+```
+
+#### On Bare Metal
+
+On bare metal, `Node.Status.Addresses` will be set by kubelet
```

Comment:
Yeah, I think the current "auto" code is not great, and probably a lot of bare metal users with multiple IPs just always manually override `--node-ip`...

But this KEP is (currently) not trying to change bare metal behavior, other than extending the current bare metal behavior to do two IPs instead of just one.

Reply:
"auto" can *never* be smart enough for all clusters. Sometimes you can't know what address to use just looking at the local node, you must have a cluster view.

Reply:
> e.g. `--node-ip-mode=DNSFQDN` vs `--node-ip-mode=LocalAddr=10.128.0.0/16`

Something like `--node-ip-mode` might also help with the compatibility issues in non-bare-metal too.

The default, initially, would be `--node-ip-mode=legacy`, which means basically the current behavior (single-stack node addresses only; built-in and external cloud-providers interpret `--node-ip` differently; bare-metal autodetects the way bare-metal has always autodetected). But you can explicitly specify `--node-ip-mode=default` to get behavior more like what's in this KEP (dual-stack node addresses where available; attempt to auto-detect dual-stack by default, with fall back to single-stack; built-in and external cloud-providers use the same rules) but we could also throw in some extra fixes (have dual-stack `Node.Status.Addresses` values even in single-stack clusters, since we know it's OK because they opted in; fix bare-metal autodetection to not assume nodename==hostname, etc).

Then in release N+1 or N+2, `--node-ip-mode=default` actually becomes the default, but `--node-ip-mode=legacy` is still available. Eventually `--node-ip-mode=legacy` would go away... or maybe not?